src/lib.rs

```diff
@@ -1 +1,109 @@
 #![no_std]
+
+use core::u32;
+
+mod convert;
+mod fmt;
+mod ops;
+#[cfg(feature = "serde")]
+mod serde;
+
+/// An index into a string.
+///
+/// The index is stored as a 32 bit integer,
+/// assuming we only deal with text shorter than 4 GiB.
+#[derive(Copy, Clone, Default, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub struct StrIndex {
+    raw: u32,
+}
+
+impl StrIndex {
+    /// Index equal to the string length of this `char`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use str_index::*;
+    /// assert_eq!(
+    ///     StrIndex::from_char_len('😂'),
+    ///     StrIndex::from(4),
+    /// );
+    /// ```
+    pub fn from_char_len(c: char) -> Self {
+        StrIndex::from(c.len_utf8() as u32)
+    }
```

Comment:
Might be worth having a utf16 version of this?

Reply:
I think the correct place to do the Rust utf8 str indexes to utf16 indexes is at the language server protocol edge. Especially since `StrIndex` is supposed to be able to index into `str`.